aws/endpoints.go

```diff
@@ -54,8 +54,27 @@ type Endpoint struct {
 	// The signing method that should be used for signing the requests to the
 	// endpoint.
 	SigningMethod string
+
+	// The source of the Endpoint. By default, this will be ServiceMetadata.
+	// When providing a custom endpoint, you should set the source as CustomSource.
+	// If source is not provided when providing a custom endpoint, the SDK may not
+	// perform required host mutations correctly. Source should be used along with
+	// HostnameImmutable property as per the usage requirement.
+	Source EndpointSource
 }
 
+// EndpointSource is the source type of endpoint.
+type EndpointSource int
+
+const (
+	// ServiceMetadata denotes service provided endpoint metadata used as Endpoint Source.
+	ServiceMetadata EndpointSource = iota
+
+	// CustomSource denotes endpoint is a custom endpoint.
+	// This source should be used when providing a custom endpoint to be used by the SDK.
+	CustomSource
+)
+
```

Comment:
Forgot to mention these probably should be prefixed With `EndpointSource` since this is in the pretty large namespace of `aws` packag.e